### PR TITLE
Implement new template API

### PR DESCRIPTION
**Pull request by deniskoronchik**
* [ ] Read PR [documentation](http://ostis-dev.github.io/sc-machine/dev/pr/)
* [ ] Update changelog
* [ ] Update documentation

Merge after: #378
Resolves: #372 
Resolves: #297
Resolves: #287
Resolves: #281
Resolves: #223

Compare new API performance to `dev` branch:

Screen Shot 2021-01-30 at 19 40 39

Did some performance improvements:

Screen Shot 2021-01-31 at 01 13 02

Next steps to improve performance: #388

### CODE REVIEW

**Review by KovalM — CHANGES_REQUESTED** [review]

**Review by ValikVityaz — CHANGES_REQUESTED** [review]

File: docs/cpp/templates.md [hunk 1]
```diff
@@ -1,173 +1,165 @@
...
+    device_addr,    // sc-addr of device node
+    ScType::EDGE_ACCESS_VAR_POS_PERM,
+    ScType::NODE_VAR >> "_device_instance")
+  .Triple(
+    device_enabled_addr,    // sc-addr of device_enabled node
+    ScType::EDGE_ACCESS_VAR_POS_PERM,
+    "_device_instance")
+  .Make();
+
+ScTemplateSeach search(ctx, *templ);
```
Comment (MikhailSadovsky) [inline docs/cpp/templates.md:111]:
```suggestion
ScTemplateSearch search(ctx, *templ);
```

File: docs/cpp/templates.md [hunk 2]
```diff
@@ -1,173 +1,165 @@
...
+Search algorithm trying to find all possible variants of specified construction where _constants_ corresponds to _variables_ from template . It use any constants (available `ScAddr`'s from parameters to find equal constructions in sc-memory).
+
+It is possible to run search with parameters:
+
+```cpp
+...
+ScTemplateParams params;
+params.Add("_device_instance", instanceAddr); // element _device_instance in template become known
+
+ScTemplateSeach search(ctx, *templ, params);
```
Comment (MikhailSadovsky) [inline docs/cpp/templates.md:129]:
```suggestion
ScTemplateSearch search(ctx, *templ, params);
```

File: docs/cpp/templates.md [hunk 3]
```diff
@@ -1,173 +1,165 @@
...
+## Template parameters
 
-Search algorithm trying to find all possible variants of specified construction. It use any constants (available `ScAddr`'s from parameters to find equal constructions in sc-memory).
+It is possible to use parameters in template search/generation. Please use `ScTemplateParams` class for that.
 
 
+## Search by template
 
-## Search in construction
+Search engine traverses cs-memory to find all contructions that correspond to template. Let's assume that we need to check if specified sc-element (`_device`) is included into `device` set and `device_enabled` set:
```
Comment (MikhailSadovsky) [inline docs/cpp/templates.md:90]:
```suggestion
Search engine traverses sc-memory to find all contructions that correspond to template. Let's assume that we need to check if specified sc-element (`_device`) is included into `device` set and `device_enabled` set:
```

**Review by MikhailSadovsky — CHANGES_REQUESTED** [review]

File: sc-memory/sc-memory/sc_template_search.hpp [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+     , m_struct(namedStruct)
+     , m_index(idx)
+    {
+    }
+
+  public:
+
+    bool operator == (Iterator const & other) const;
+    bool operator != (Iterator const & other) const;
+    Iterator & operator ++ ();
```
[the remark this answers is not in the document]
Comment (deniskoronchik) [reply sc-memory/sc-memory/sc_template_search.hpp:39]: #391

File: sc-memory/sc-memory/sc_template_builder.hpp [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  _SC_EXTERN ScTemplateBuilder & TripleWithRelation(ScTemplateArg const & source,
+                                                    ScTemplateArg const & edge,
+                                                    ScTemplateArg const & target,
+                                                    ScTemplateArg const & attrEdge,
+                                                    ScTemplateArg const & attr);
+
+  /*! \brief Makes template based on appended triples
+   *  \return Returns pointer to created template
+   */
+  _SC_EXTERN ScTemplatePtr Make();
```
[the remark this answers is not in the document]
Comment (deniskoronchik) [reply sc-memory/sc-memory/sc_template_builder.hpp:48]:
```cpp
ScTemplateBuilder builder;

builder.build(); // ?
```

File: sc-memory/sc-memory/sc_template_builder_impl.hpp [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+   * possible abuse, use result name mapping, and get result by names
+   */
+  ScTemplateBuilderImpl & TripleWithRelation(ScTemplateArg const & param1,
+                                             ScTemplateArg const & param2,
+                                             ScTemplateArg const & param3,
+                                             ScTemplateArg const & param4,
+                                             ScTemplateArg const & param5);
+
+  /*! Returns pointer to created template
+  * \remarks Do not use builde after this method call.
```
[the remark this answers is not in the document]
Comment (deniskoronchik) [reply sc-memory/sc-memory/sc_template_builder_impl.hpp:36]: Fixed to: `* \remarks Do not use builder after this method call.`

This comment talks to do not use builder (call any function) after this functions call.

File: sc-memory/sc-memory/sc_template_builder.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (KovalM) [reply sc-memory/sc-memory/sc_template_builder.hpp:48]: Yes

**Review by KovalM — APPROVED** [review]

**Review by MikhailSadovsky — APPROVED** [review]

File: sc-memory/sc-memory/sc_template_builder.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (deniskoronchik) [reply sc-memory/sc-memory/sc_template_builder.hpp:48]: We can discuss that later. I don't see any problems with current name. And don't understand why new one will be better. The main idea of this name was, that building of template is going during `Tirple...` method calls. These methods can raise an exception. And `Make` method produce a final pointer to builded template.

Also it was similar to make_unique, make_shared in std library

[the remark this answers is not in the document]
Comment (KovalM) [reply sc-memory/sc-memory/sc_template_builder.hpp:48]: My single argument was popularity of the method's name "build" among developers, but I investigate cpp code from different projects and found that is not so popular as I thought. This name so popular among Java developers. I am ready to resolve this conversation.

**Review by ValikVityaz — APPROVED** [review]

**Pull request merged**